Github Action to automatically sync content

stp-ip opened the pull request:
Content is synced from coredns/coredns and checked before being
committed into the same repository so master should always reflect the
correct state.

It is using a scheduled github action for now as this removes any github
authentication coordination and simplifies the general action.

This prepares further automation for Netlify.

Being run every 12 minutes starting on the 4th minute.

Note to reviewer:
We currently push changes even so corecheck fails. We should analyse on, which it fails and maybe fix or rewrite them to not use "corefile" as syntax to remove checking.

stp-ip commented:
We could move this to be run on a dispatch event that would be run during release or similar. Basically sending a POST request with a personal github token.

Wanted to be simple first.

miekg commented on .github/workflows/sync.coredns.yml:
```diff
@@ -0,0 +1,54 @@
+          echo "::add-path::$(go env GOPATH)/bin"
+        shell: bash
+      -
+        name: Deps
```
why do need we need to build coredns?

stp-ip replied:
The CoreDNS binary is needed by corecheck afaik.
As we already have the source code building seemed like the easiest solution especially considering we are checking the corefiles within the documentation of that specific commit of coredns anyway

miekg replied:
hmm, you're right; seems to be heavy weight to just pushing to netifly; we better leave this out and do it off line (for now). Also if coredns doesn't compile we can't deploy (just mentioning this)

stp-ip replied:
Deploy on netlify is separate from this.
I'll push a new version, where the corefile check is done in a separate action. That way we have an idea on issues even so we don't cancel a sync etc.

miekg commented on Makefile:
```diff
@@ -1,25 +1,15 @@
-clean:
-	rm -rf /var/www/coredns.io/*
-
-.PHONY: test
```
this might be still useful?

stp-ip replied:
Happy to keep something along the lines of `make server`. Will push that.

stp-ip replied:
Kept it as `make run` see latest commit.

miekg commented:
Note that make sync *always* changes the date on the plugin's metadata; we may want to change that as well in the python code; use the current month or something, not the current time otherwise we will always be pushing changes out

stp-ip commented:
Will check the time update.

stp-ip commented:
* Split up corefile test and sync.
* Replaced syncing python code with Go (tried with an AST solution, but simplified it back)
  * Syncing now takes into account content changes and only overwrites, if necessary.
  * Syncing now uses a toml lib to reduce manual string constructs
* Fixed merge conflicts.

stp-ip commented:
Would love another review @miekg

miekg commented:
[ Quoting <notifications@github.com> in "Re: [coredns/coredns.io] Github Act..." ]
>Would love another review @miekg

oh yeah, sorry. On the TODO. Think it lgtm, and we can just go forward.

Do we have a domain name pointing to it?

stp-ip commented:
On the netlify side it's currently netlify.coredns.io
Once we feel comfortable with both the auto sync and netlify auto deploy we can switch www.coredns.io or about.coredns.io it. (Or use the apex as we currently do)

We could then redirect coredns.io to the new subdomain.

stp-ip commented:
* coredns.netlify.com

stp-ip commented:
Doing a follow up PR to fix the go script. Seems to update time even so it shouldn't.
